File: dotnet/src/SodiumFRP/FSharp/SodiumFRP.FSharp/MutableListener.fs
module SodiumFRP.MutableListener

let setListener l (m : MutableListener) = m.SetListenerImpl l
let clearListener (m : MutableListener) = m.ClearListenerImpl ()
let unlisten (m : MutableListener) = Listener.unlisten m
let getListenerWithWeakReference (m : MutableListener) = Listener.getListenerWithWeakReference m

File: dotnet/src/Sodium/Sodium.Core.Frp/MutableListener.cs
namespace Sodium.Frp
{
    public class MutableListener : IListener
    {
        private readonly WeakMutableListener weakMutableListener = new WeakMutableListener();
        private IListener listener;

        internal void SetListenerImpl(IListener listener)
        {
            this.listener = listener;
            this.weakMutableListener.WeakListener = listener?.GetListenerWithWeakReference();
        }

        internal void ClearListenerImpl()
        {
            this.listener = null;
            this.weakMutableListener.WeakListener = null;
        }

        void IListener.Unlisten() => this.listener?.Unlisten();
        IListenerWithWeakReference IListener.GetListenerWithWeakReference() => this.weakMutableListener;

        private class WeakMutableListener : IListenerWithWeakReference
        {
            public IListenerWithWeakReference WeakListener;

            void IListenerWithWeakReference.Unlisten() => this.WeakListener?.Unlisten();
        }
    }
}